dvc/remote/base.py

```diff
@@ -505,17 +507,37 @@ def pull(self, cache, named_cache, jobs=None, show_checksums=False):
 
         return ret
 
-    def transfer(self, from_fs, from_info, jobs=None, no_progress_bar=False):
-        from dvc.objects.transfer import transfer
-
-        return transfer(
+    def transfer(
+        self,
+        from_fs,
+        from_info,
+        jobs=None,
+        update=False,
+        no_progress_bar=False,
+    ):
+        # When running import-url --to-remote / add --to-remote/-o ... we
+        # assume that it is unlikely that the odb will contain majority of the
+        # hashes, so we transfer everything as is (even if that file might
+        # already be in the cache) and don't waste an upload to scan the layout
+        # of the source location. But when doing update --to-remote, there is
+        # a high probability that the odb might contain some of the hashes, so
+        # we first calculate all the hashes (but don't transfer anything) and
+        # then only update the missing cache files.
+
+        upload = not (update and from_fs.isdir(from_info))
+        obj = stage(
             self.odb,
-            from_fs,
             from_info,
+            from_fs,
+            "md5",
+            upload=upload,
             jobs=jobs,
             no_progress_bar=no_progress_bar,
         )
 
+        save(self.odb, obj)
```

Comment:
One thing that worries me is that, when `save()` calls `odb.add` with an `fs` of the same type (like both `fs` and `odb.fs` are on s3) then it `move()`s the file instead of `copy()` which contradicts with it's second behavior (just `upload_fobj`ing but not removing) so I guess we might need to introduce an option like `use_copy=...` to `save()` and `odb.add()` and use `self.fs.copy()` instead;

https://github.com/iterative/dvc/blob/2472c04873305fdfae5e93a0201d3af0d4689702/dvc/objects/db/base.py#L55-L59

CC: @efiop

Reply:
@isidentical Good point! The purpose of that is to optimize for storage (e.g. `dvc add bigdatafile), but it indeed should be an explicit strategy option. Finding a good name for it is not easy though... Maybe `move=True/False` instead?